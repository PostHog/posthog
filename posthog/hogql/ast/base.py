import re
from typing import Optional, Literal

from pydantic import BaseModel, Extra

from posthog.hogql.constants import ConstantDataType
from posthog.hogql.errors import NotImplementedException

# NOTE: when you add new AST fields or nodes, add them to the Visitor classes in visitor.py as well!

camel_case_pattern = re.compile(r"(?<!^)(?=[A-Z])")


class AST(BaseModel):
    class Config:
        extra = Extra.forbid

    def accept(self, visitor):
        camel_case_name = camel_case_pattern.sub("_", self.__class__.__name__).lower()
        method_name = f"visit_{camel_case_name}"
        if hasattr(visitor, method_name):
            visit = getattr(visitor, method_name)
            return visit(self)
        if hasattr(visitor, "visit_unknown"):
            return visitor.visit_unknown(self)
        raise NotImplementedException(f"Visitor has no method {method_name}")


class Type(AST):
    def get_child(self, name: str) -> "Type":
        raise NotImplementedException("Type.get_child not overridden")

    def has_child(self, name: str) -> bool:
        return self.get_child(name) is not None

    def resolve_constant_type(self) -> Optional["ConstantType"]:
        from posthog.hogql.ast import UnknownType

        return UnknownType()


class ConstantType(Type):
    data_type: ConstantDataType

    def resolve_constant_type(self) -> "ConstantType":
        return self


class Expr(AST):
    type: Optional[Type]


class Macro(Expr):
    name: str
    expr: Expr
    # Whether the macro is an inlined column "SELECT 1 AS a" or a subquery "SELECT a AS (SELECT 1)"
    macro_format: Literal["column", "subquery"]
