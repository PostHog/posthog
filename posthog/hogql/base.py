import re
from dataclasses import dataclass, field

from typing import Literal, Optional

from posthog.hogql.constants import ConstantDataType
from posthog.hogql.errors import NotImplementedException


# Given a string like "CorrectHorseBS", match the "H" and "B", so that we can convert this to "correct_horse_bs"
camel_case_pattern = re.compile(r"(?<!^)(?<![A-Z])(?=[A-Z])")


@dataclass(kw_only=True)
class AST:
    start: Optional[int] = field(default=None)
    end: Optional[int] = field(default=None)

    def accept(self, visitor):
        camel_case_name = camel_case_pattern.sub("_", self.__class__.__name__).lower()
        if "hog_qlx" in camel_case_name:
            camel_case_name = camel_case_name.replace("hog_qlx", "hogqlx_")
        method_name = f"visit_{camel_case_name}"
        if hasattr(visitor, method_name):
            visit = getattr(visitor, method_name)
            return visit(self)
        if hasattr(visitor, "visit_unknown"):
            return visitor.visit_unknown(self)
        raise NotImplementedException(f"Visitor has no method {method_name}")


@dataclass(kw_only=True)
class Type(AST):
    def get_child(self, name: str) -> "Type":
        raise NotImplementedException("Type.get_child not overridden")

    def has_child(self, name: str) -> bool:
        return self.get_child(name) is not None

    def resolve_constant_type(self) -> Optional["ConstantType"]:
        return UnknownType()


@dataclass(kw_only=True)
class Expr(AST):
    type: Optional[Type] = field(default=None)


@dataclass(kw_only=True)
class CTE(Expr):
    """A common table expression."""

    name: str
    expr: Expr
    # Whether the CTE is an inlined column "WITH 1 AS a" or a subquery "WITH a AS (SELECT 1)"
    cte_type: Literal["column", "subquery"]


@dataclass(kw_only=True)
class ConstantType(Type):
    data_type: ConstantDataType

    def resolve_constant_type(self) -> "ConstantType":
        return self

    def print_type(self) -> str:
        raise NotImplementedException("ConstantType.print_type not implemented")


@dataclass(kw_only=True)
class UnknownType(ConstantType):
    data_type: ConstantDataType = field(default="unknown", init=False)

    def print_type(self) -> str:
        return "Unknown"
