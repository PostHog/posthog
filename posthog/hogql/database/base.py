from typing import Any, Callable, Dict, List

from pydantic import BaseModel, Extra

from posthog.hogql.errors import HogQLException, NotImplementedException
from posthog.hogql.database.fields import DatabaseField


class Table(BaseModel):
    class Config:
        extra = Extra.forbid

    def has_field(self, name: str) -> bool:
        return hasattr(self, name)

    def get_field(self, name: str) -> DatabaseField:
        if self.has_field(name):
            return getattr(self, name)
        raise HogQLException(f'Field "{name}" not found on table {self.__class__.__name__}')

    def clickhouse_table(self):
        raise NotImplementedException("Table.clickhouse_table not overridden")

    def hogql_table(self):
        raise NotImplementedException("Table.hogql_table not overridden")

    def avoid_asterisk_fields(self) -> List[str]:
        return []

    def get_asterisk(self) -> Dict[str, DatabaseField]:
        asterisk: Dict[str, DatabaseField] = {}
        fields_to_avoid = self.avoid_asterisk_fields() + ["team_id"]
        for key, field in self.__fields__.items():
            if key in fields_to_avoid:
                continue
            database_field = field.default
            if isinstance(database_field, DatabaseField):
                asterisk[key] = database_field
            elif (
                isinstance(database_field, Table)
                or isinstance(database_field, LazyJoin)
                or isinstance(database_field, FieldTraverser)
            ):
                pass  # ignore virtual tables for now
            else:
                raise HogQLException(f"Unknown field type {type(database_field).__name__} for asterisk")
        return asterisk


class LazyJoin(BaseModel):
    class Config:
        extra = Extra.forbid

    join_function: Callable[[str, str, Dict[str, Any]], Any]
    join_table: Table
    from_field: str


class LazyTable(Table):
    class Config:
        extra = Extra.forbid

    def lazy_select(self, requested_fields: Dict[str, Any]) -> Any:
        raise NotImplementedException("LazyTable.lazy_select not overridden")


class VirtualTable(Table):
    class Config:
        extra = Extra.forbid


class FieldTraverser(BaseModel):
    class Config:
        extra = Extra.forbid

    chain: List[str]
