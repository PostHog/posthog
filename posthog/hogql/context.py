from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Dict, List, Literal, Optional, Any

from posthog.hogql.timings import HogQLTimings
from posthog.schema import HogQLNotice, HogQLQueryModifiers

if TYPE_CHECKING:
    from posthog.hogql.database.database import Database


@dataclass
class HogQLFieldAccess:
    input: List[str]
    type: Optional[Literal["event", "event.properties", "person", "person.properties"]]
    field: Optional[str]
    sql: str


@dataclass
class HogQLContext:
    """Context given to a HogQL expression printer"""

    # Team making the queries
    team_id: Optional[int]
    # Virtual database we're querying, will be populated from team_id if not present
    database: Optional["Database"] = None
    # If set, will save string constants to this dict. Inlines strings into the query if None.
    values: Dict = field(default_factory=dict)
    # Are we small part of a non-HogQL query? If so, use custom syntax for accessed person properties.
    within_non_hogql_query: bool = False
    # Enable full SELECT queries and subqueries in ClickHouse
    enable_select_queries: bool = False
    # Do we apply a limit of MAX_SELECT_RETURNED_ROWS=10000 to the topmost select query?
    limit_top_select: bool = True
    # How many nested views do we support on this query?
    max_view_depth: int = 1

    # Warnings returned with the metadata query
    warnings: List["HogQLNotice"] = field(default_factory=list)
    # Notices returned with the metadata query
    notices: List["HogQLNotice"] = field(default_factory=list)
    # Timings in seconds for different parts of the HogQL query
    timings: HogQLTimings = field(default_factory=HogQLTimings)
    # Modifications requested by the HogQL client
    modifiers: HogQLQueryModifiers = field(default_factory=HogQLQueryModifiers)

    def add_value(self, value: Any) -> str:
        key = f"hogql_val_{len(self.values)}"
        self.values[key] = value
        return f"%({key})s"

    def add_sensitive_value(self, value: Any) -> str:
        key = f"hogql_val_{len(self.values)}_sensitive"
        self.values[key] = value
        return f"%({key})s"

    def add_notice(
        self,
        message: str,
        start: Optional[int] = None,
        end: Optional[int] = None,
        fix: Optional[str] = None,
    ):
        if not any(n.start == start and n.end == end and n.message == message and n.fix == fix for n in self.notices):
            self.notices.append(HogQLNotice(start=start, end=end, message=message, fix=fix))
