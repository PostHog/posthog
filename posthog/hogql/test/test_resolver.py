from datetime import timezone, datetime, date
from typing import Optional, Dict, cast
import pytest
from django.test import override_settings
from uuid import UUID

from freezegun import freeze_time

from posthog.hogql import ast
from posthog.hogql.context import HogQLContext
from posthog.hogql.database.database import create_hogql_database
from posthog.hogql.database.models import (
    FieldTraverser,
    StringJSONDatabaseField,
    StringDatabaseField,
    DateTimeDatabaseField,
)
from posthog.hogql.test.utils import pretty_dataclasses
from posthog.hogql.visitor import clone_expr
from posthog.hogql.parser import parse_select
from posthog.hogql.printer import print_ast, print_prepared_ast
from posthog.hogql.resolver import ResolverException, resolve_types
from posthog.test.base import BaseTest


class TestResolver(BaseTest):
    maxDiff = None

    def _select(self, query: str, placeholders: Optional[Dict[str, ast.Expr]] = None) -> ast.SelectQuery:
        return cast(
            ast.SelectQuery,
            clone_expr(parse_select(query, placeholders=placeholders), clear_locations=True),
        )

    def _print_hogql(self, select: str):
        expr = self._select(select)
        return print_ast(
            expr,
            HogQLContext(team_id=self.team.pk, enable_select_queries=True),
            "hogql",
        )

    def setUp(self):
        self.database = create_hogql_database(self.team.pk)
        self.context = HogQLContext(database=self.database, team_id=self.team.pk)

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_resolve_events_table(self):
        expr = self._select("SELECT event, events.timestamp FROM events WHERE events.event = 'test'")
        expr = resolve_types(expr, self.context, dialect="clickhouse")
        assert pretty_dataclasses(expr) == self.snapshot

    def test_will_not_run_twice(self):
        expr = self._select("SELECT event, events.timestamp FROM events WHERE events.event = 'test'")
        expr = resolve_types(expr, self.context, dialect="clickhouse")
        with self.assertRaises(ResolverException) as context:
            expr = resolve_types(expr, self.context, dialect="clickhouse")
        self.assertEqual(
            str(context.exception),
            "Type already resolved for SelectQuery (SelectQueryType). Can't run again.",
        )

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_resolve_events_table_alias(self):
        expr = self._select("SELECT event, e.timestamp FROM events e WHERE e.event = 'test'")
        expr = resolve_types(expr, self.context, dialect="clickhouse")
        assert pretty_dataclasses(expr) == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_resolve_events_table_column_alias(self):
        expr = self._select("SELECT event as ee, ee, ee as e, e.timestamp FROM events e WHERE e.event = 'test'")
        expr = resolve_types(expr, self.context, dialect="clickhouse")
        assert pretty_dataclasses(expr) == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_resolve_events_table_column_alias_inside_subquery(self):
        expr = self._select("SELECT b FROM (select event as b, timestamp as c from events) e WHERE e.b = 'test'")
        expr = resolve_types(expr, self.context, dialect="clickhouse")
        assert pretty_dataclasses(expr) == self.snapshot

    def test_resolve_subquery_no_field_access(self):
        # From ClickHouse's GitHub: "Aliases defined outside of subquery are not visible in subqueries (but see below)."
        expr = self._select(
            "SELECT event, (select count() from events where event = e.event) as c FROM events e where event = '$pageview'"
        )
        with self.assertRaises(ResolverException) as e:
            expr = resolve_types(expr, self.context, dialect="clickhouse")
        self.assertEqual(str(e.exception), "Unable to resolve field: e")

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_resolve_constant_type(self):
        with freeze_time("2020-01-10 00:00:00"):
            expr = self._select(
                "SELECT 1, 'boo', true, 1.1232, null, {date}, {datetime}, {uuid}, {array}, {array12}, {tuple}",
                placeholders={
                    "date": ast.Constant(value=date(2020, 1, 10)),
                    "datetime": ast.Constant(value=datetime(2020, 1, 10, 0, 0, 0, tzinfo=timezone.utc)),
                    "uuid": ast.Constant(value=UUID("00000000-0000-4000-8000-000000000000")),
                    "array": ast.Constant(value=[]),
                    "array12": ast.Constant(value=[1, 2]),
                    "tuple": ast.Constant(value=(1, 2, 3)),
                },
            )
            expr = resolve_types(expr, self.context, dialect="clickhouse")
            assert pretty_dataclasses(expr) == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_resolve_boolean_operation_types(self):
        expr = self._select("SELECT 1 and 1, 1 or 1, not true")
        expr = resolve_types(expr, self.context, dialect="clickhouse")
        assert pretty_dataclasses(expr) == self.snapshot

    def test_resolve_errors(self):
        queries = [
            "SELECT event, (select count() from events where event = x.event) as c FROM events x where event = '$pageview'",
            "SELECT x, (SELECT 1 AS x)",
            "SELECT x IN (SELECT 1 AS x)",
            "SELECT events.x FROM (SELECT event as x FROM events) AS t",
            "SELECT x.y FROM (SELECT event as y FROM events AS x) AS t",
        ]
        for query in queries:
            with self.assertRaises(ResolverException) as e:
                resolve_types(self._select(query), self.context, dialect="clickhouse")
            self.assertIn("Unable to resolve field:", str(e.exception))

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_resolve_lazy_pdi_person_table(self):
        expr = self._select("select distinct_id, person.id from person_distinct_ids")
        expr = resolve_types(expr, self.context, dialect="clickhouse")
        assert pretty_dataclasses(expr) == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_resolve_lazy_events_pdi_table(self):
        expr = self._select("select event, pdi.person_id from events")
        expr = resolve_types(expr, self.context, dialect="clickhouse")
        assert pretty_dataclasses(expr) == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_resolve_lazy_events_pdi_table_aliased(self):
        expr = self._select("select event, e.pdi.person_id from events e")
        expr = resolve_types(expr, self.context, dialect="clickhouse")
        assert pretty_dataclasses(expr) == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_resolve_lazy_events_pdi_person_table(self):
        expr = self._select("select event, pdi.person.id from events")
        expr = resolve_types(expr, self.context, dialect="clickhouse")
        assert pretty_dataclasses(expr) == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_resolve_lazy_events_pdi_person_table_aliased(self):
        expr = self._select("select event, e.pdi.person.id from events e")
        expr = resolve_types(expr, self.context, dialect="clickhouse")
        assert pretty_dataclasses(expr) == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_resolve_virtual_events_poe(self):
        expr = self._select("select event, poe.id from events")
        expr = resolve_types(expr, self.context, dialect="clickhouse")
        assert pretty_dataclasses(expr) == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_resolve_union_all(self):
        node = self._select("select event, timestamp from events union all select event, timestamp from events")
        node = resolve_types(node, self.context, dialect="clickhouse")
        assert pretty_dataclasses(node) == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_call_type(self):
        node = self._select("select max(timestamp) from events")
        node = resolve_types(node, self.context, dialect="clickhouse")
        assert pretty_dataclasses(node) == self.snapshot

    def test_ctes_loop(self):
        with self.assertRaises(ResolverException) as e:
            self._print_hogql("with cte as (select * from cte) select * from cte")
        self.assertIn("Too many CTE expansions (50+). Probably a CTE loop.", str(e.exception))

    def test_ctes_basic_column(self):
        expr = self._print_hogql("with 1 as cte select cte from events")
        expected = self._print_hogql("select 1 from events")
        self.assertEqual(expr, expected)

    def test_ctes_recursive_column(self):
        self.assertEqual(
            self._print_hogql("with 1 as cte, cte as soap select soap from events"),
            self._print_hogql("select 1 from events"),
        )

    def test_ctes_field_access(self):
        with self.assertRaises(ResolverException) as e:
            self._print_hogql("with properties as cte select cte.$browser from events")
        self.assertIn("Cannot access fields on CTE cte yet", str(e.exception))

    def test_ctes_subqueries(self):
        self.assertEqual(
            self._print_hogql("with my_table as (select * from events) select * from my_table"),
            self._print_hogql("select * from (select * from events) my_table"),
        )

        self.assertEqual(
            self._print_hogql("with my_table as (select * from events) select my_table.timestamp from my_table"),
            self._print_hogql("select my_table.timestamp from (select * from events) my_table"),
        )

        self.assertEqual(
            self._print_hogql("with my_table as (select * from events) select timestamp from my_table"),
            self._print_hogql("select timestamp from (select * from events) my_table"),
        )

    def test_ctes_subquery_deep(self):
        self.assertEqual(
            self._print_hogql(
                "with my_table as (select * from events), "
                "other_table as (select * from (select * from (select * from my_table))) "
                "select * from other_table"
            ),
            self._print_hogql(
                "select * from (select * from (select * from (select * from (select * from events) as my_table))) as other_table"
            ),
        )

    def test_ctes_subquery_recursion(self):
        self.assertEqual(
            self._print_hogql(
                "with users as (select event, timestamp as tt from events ), final as ( select tt from users ) select * from final"
            ),
            self._print_hogql(
                "select * from (select tt from (select event, timestamp as tt from events) AS users) AS final"
            ),
        )

    @override_settings(PERSON_ON_EVENTS_OVERRIDE=False, PERSON_ON_EVENTS_V2_OVERRIDE=False)
    @pytest.mark.usefixtures("unittest_snapshot")
    def test_asterisk_expander_table(self):
        self.setUp()  # rebuild self.database with PERSON_ON_EVENTS_OVERRIDE=False
        node = self._select("select * from events")
        node = resolve_types(node, self.context, dialect="clickhouse")
        assert pretty_dataclasses(node) == self.snapshot

    @override_settings(PERSON_ON_EVENTS_OVERRIDE=False, PERSON_ON_EVENTS_V2_OVERRIDE=False)
    @pytest.mark.usefixtures("unittest_snapshot")
    def test_asterisk_expander_table_alias(self):
        self.setUp()  # rebuild self.database with PERSON_ON_EVENTS_OVERRIDE=False
        node = self._select("select * from events e")
        node = resolve_types(node, self.context, dialect="clickhouse")
        assert pretty_dataclasses(node) == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_asterisk_expander_subquery(self):
        node = self._select("select * from (select 1 as a, 2 as b)")
        node = resolve_types(node, self.context, dialect="clickhouse")
        assert pretty_dataclasses(node) == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_asterisk_expander_subquery_alias(self):
        node = self._select("select x.* from (select 1 as a, 2 as b) x")
        node = resolve_types(node, self.context, dialect="clickhouse")
        assert pretty_dataclasses(node) == self.snapshot

    @override_settings(PERSON_ON_EVENTS_OVERRIDE=False, PERSON_ON_EVENTS_V2_OVERRIDE=False)
    @pytest.mark.usefixtures("unittest_snapshot")
    def test_asterisk_expander_from_subquery_table(self):
        self.setUp()  # rebuild self.database with PERSON_ON_EVENTS_OVERRIDE=False
        node = self._select("select * from (select * from events)")
        node = resolve_types(node, self.context, dialect="clickhouse")
        assert pretty_dataclasses(node) == self.snapshot

    def test_asterisk_expander_multiple_table_error(self):
        node = self._select("select * from (select 1 as a, 2 as b) x left join (select 1 as a, 2 as b) y on x.a = y.a")
        with self.assertRaises(ResolverException) as e:
            resolve_types(node, self.context, dialect="clickhouse")
        self.assertEqual(
            str(e.exception),
            "Cannot use '*' without table name when there are multiple tables in the query",
        )

    @override_settings(PERSON_ON_EVENTS_OVERRIDE=False, PERSON_ON_EVENTS_V2_OVERRIDE=False)
    @pytest.mark.usefixtures("unittest_snapshot")
    def test_asterisk_expander_select_union(self):
        self.setUp()  # rebuild self.database with PERSON_ON_EVENTS_OVERRIDE=False
        node = self._select("select * from (select * from events union all select * from events)")
        node = resolve_types(node, self.context, dialect="clickhouse")
        assert pretty_dataclasses(node) == self.snapshot

    def test_lambda_parent_scope(self):
        # does not raise
        node = self._select("select timestamp, arrayMap(x -> x + timestamp, [2]) from events")
        node = cast(ast.SelectQuery, resolve_types(node, self.context, dialect="clickhouse"))

        # found a type
        lambda_type: ast.SelectQueryType = cast(ast.SelectQueryType, cast(ast.Call, node.select[1]).args[0].type)
        self.assertEqual(lambda_type.parent, node.type)
        self.assertEqual(list(lambda_type.aliases.keys()), ["x"])
        self.assertEqual(list(lambda_type.parent.columns.keys()), ["timestamp"])

    def test_field_traverser_double_dot(self):
        # Create a condition where we want to ".." out of "events.poe." to get to a higher level prop
        self.database.events.fields["person"] = FieldTraverser(chain=["poe"])
        self.database.events.fields["poe"].fields["id"] = FieldTraverser(chain=["..", "pdi", "person_id"])
        self.database.events.fields["poe"].fields["created_at"] = FieldTraverser(
            chain=["..", "pdi", "person", "created_at"]
        )
        self.database.events.fields["poe"].fields["properties"] = StringJSONDatabaseField(name="person_properties")

        node = self._select("SELECT event, person.id, person.properties, person.created_at FROM events")
        node = cast(ast.SelectQuery, resolve_types(node, self.context, dialect="clickhouse"))

        # all columns resolve to a type in the end
        assert cast(ast.FieldType, node.select[0].type).resolve_database_field() == StringDatabaseField(
            name="event", array=None, nullable=None
        )
        assert cast(ast.FieldType, node.select[1].type).resolve_database_field() == StringDatabaseField(
            name="person_id", array=None, nullable=None
        )
        assert cast(ast.FieldType, node.select[2].type).resolve_database_field() == StringJSONDatabaseField(
            name="person_properties"
        )
        assert cast(ast.FieldType, node.select[3].type).resolve_database_field() == DateTimeDatabaseField(
            name="created_at", array=None, nullable=None
        )

    def test_visit_hogqlx_tag(self):
        node = self._select("select event from <HogQLQuery query='select event from events' />")
        node = cast(ast.SelectQuery, resolve_types(node, self.context, dialect="clickhouse"))
        table_node = cast(ast.SelectQuery, node).select_from.table
        expected = ast.SelectQuery(
            select=[ast.Alias(hidden=True, alias="event", expr=ast.Field(chain=["event"]))],
            select_from=ast.JoinExpr(table=ast.Field(chain=["events"])),
        )
        assert clone_expr(table_node, clear_types=True) == expected

    def test_visit_hogqlx_tag_alias(self):
        node = self._select("select event from <HogQLQuery query='select event from events' /> a")
        node = cast(ast.SelectQuery, resolve_types(node, self.context, dialect="clickhouse"))
        assert cast(ast.SelectQuery, node).select_from.alias == "a"

    def test_visit_hogqlx_tag_source(self):
        query = """
            select id, email from (
                <PersonsQuery
                    select={['id', 'properties.email as email']}
                    source={
                        <HogQLQuery query='select distinct person_id from events' />
                    }
                />
            )
        """
        node = cast(ast.SelectQuery, resolve_types(self._select(query), self.context, dialect="hogql"))
        hogql = print_prepared_ast(node, HogQLContext(team_id=self.team.pk, enable_select_queries=True), "hogql")
        expected = (
            f"SELECT id, email FROM "
            f"(SELECT id, properties.email AS email FROM persons WHERE in(id, "
            f"(SELECT DISTINCT person_id FROM events)"
            f") ORDER BY id ASC LIMIT 101 OFFSET 0) "
            f"LIMIT 10000"
        )
        assert hogql == expected
