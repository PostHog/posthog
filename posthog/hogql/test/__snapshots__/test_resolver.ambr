# name: TestResolver.test_asterisk_expander_from_subquery_table
  '''
  {
    select: [
      {
        alias: "uuid"
        expr: {
          chain: [
            "uuid"
          ]
          type: {
            name: "uuid"
            table_type: {
              aliases: {}
              anonymous_tables: []
              columns: {
                $group_0: {
                  alias: "$group_0"
                  type: {
                    name: "$group_0"
                    table_type: {
                      table: {
                        fields: {
                          $group_0: {},
                          $group_1: {},
                          $group_2: {},
                          $group_3: {},
                          $group_4: {},
                          $session_id: {},
                          created_at: {},
                          distinct_id: {},
                          elements_chain: {},
                          event: {},
                          goe_0: {},
                          goe_1: {},
                          goe_2: {},
                          goe_3: {},
                          goe_4: {},
                          group_0: {},
                          group_1: {},
                          group_2: {},
                          group_3: {},
                          group_4: {},
                          pdi: {},
                          person: {},
                          person_id: {},
                          poe: {},
                          properties: {},
                          session: {},
                          team_id: {},
                          timestamp: {},
                          uuid: {}
                        }
                      }
                    }
                  }
                },
                $group_1: {
                  alias: "$group_1"
                  type: {
                    name: "$group_1"
                    table_type: <recursion ...>
                  }
                },
                $group_2: {
                  alias: "$group_2"
                  type: {
                    name: "$group_2"
                    table_type: <recursion ...>
                  }
                },
                $group_3: {
                  alias: "$group_3"
                  type: {
                    name: "$group_3"
                    table_type: <recursion ...>
                  }
                },
                $group_4: {
                  alias: "$group_4"
                  type: {
                    name: "$group_4"
                    table_type: <recursion ...>
                  }
                },
                $session_id: {
                  alias: "$session_id"
                  type: {
                    name: "$session_id"
                    table_type: <recursion ...>
                  }
                },
                created_at: {
                  alias: "created_at"
                  type: {
                    name: "created_at"
                    table_type: <recursion ...>
                  }
                },
                distinct_id: {
                  alias: "distinct_id"
                  type: {
                    name: "distinct_id"
                    table_type: <recursion ...>
                  }
                },
                elements_chain: {
                  alias: "elements_chain"
                  type: {
                    name: "elements_chain"
                    table_type: <recursion ...>
                  }
                },
                event: {
                  alias: "event"
                  type: {
                    name: "event"
                    table_type: <recursion ...>
                  }
                },
                properties: {
                  alias: "properties"
                  type: {
                    name: "properties"
                    table_type: <recursion ...>
                  }
                },
                timestamp: {
                  alias: "timestamp"
                  type: {
                    name: "timestamp"
                    table_type: <recursion ...>
                  }
                },
                uuid: {
                  alias: "uuid"
                  type: {
                    name: "uuid"
                    table_type: <recursion ...>
                  }
                }
              }
              ctes: {}
              tables: {
                events: <recursion ...>
              }
            }
          }
        }
        hidden: True
        type: {
          alias: "uuid"
          type: <recursion ...>
        }
      },
      {
        alias: "event"
        expr: {
          chain: [
            "event"
          ]
          type: {
            name: "event"
            table_type: <recursion ...>
          }
        }
        hidden: True
        type: {
          alias: "event"
          type: <recursion ...>
        }
      },
      {
        alias: "properties"
        expr: {
          chain: [
            "properties"
          ]
          type: {
            name: "properties"
            table_type: <recursion ...>
          }
        }
        hidden: True
        type: {
          alias: "properties"
          type: <recursion ...>
        }
      },
      {
        alias: "timestamp"
        expr: {
          chain: [
            "timestamp"
          ]
          type: {
            name: "timestamp"
            table_type: <recursion ...>
          }
        }
        hidden: True
        type: {
          alias: "timestamp"
          type: <recursion ...>
        }
      },
      {
        alias: "distinct_id"
        expr: {
          chain: [
            "distinct_id"
          ]
          type: {
            name: "distinct_id"
            table_type: <recursion ...>
          }
        }
        hidden: True
        type: {
          alias: "distinct_id"
          type: <recursion ...>
        }
      },
      {
        alias: "elements_chain"
        expr: {
          chain: [
            "elements_chain"
          ]
          type: {
            name: "elements_chain"
            table_type: <recursion ...>
          }
        }
        hidden: True
        type: {
          alias: "elements_chain"
          type: <recursion ...>
        }
      },
      {
        alias: "created_at"
        expr: {
          chain: [
            "created_at"
          ]
          type: {
            name: "created_at"
            table_type: <recursion ...>
          }
        }
        hidden: True
        type: {
          alias: "created_at"
          type: <recursion ...>
        }
      },
      {
        alias: "$session_id"
        expr: {
          chain: [
            "$session_id"
          ]
          type: {
            name: "$session_id"
            table_type: <recursion ...>
          }
        }
        hidden: True
        type: {
          alias: "$session_id"
          type: <recursion ...>
        }
      },
      {
        alias: "$group_0"
        expr: {
          chain: [
            "$group_0"
          ]
          type: {
            name: "$group_0"
            table_type: <recursion ...>
          }
        }
        hidden: True
        type: {
          alias: "$group_0"
          type: <recursion ...>
        }
      },
      {
        alias: "$group_1"
        expr: {
          chain: [
            "$group_1"
          ]
          type: {
            name: "$group_1"
            table_type: <recursion ...>
          }
        }
        hidden: True
        type: {
          alias: "$group_1"
          type: <recursion ...>
        }
      },
      {
        alias: "$group_2"
        expr: {
          chain: [
            "$group_2"
          ]
          type: {
            name: "$group_2"
            table_type: <recursion ...>
          }
        }
        hidden: True
        type: {
          alias: "$group_2"
          type: <recursion ...>
        }
      },
      {
        alias: "$group_3"
        expr: {
          chain: [
            "$group_3"
          ]
          type: {
            name: "$group_3"
            table_type: <recursion ...>
          }
        }
        hidden: True
        type: {
          alias: "$group_3"
          type: <recursion ...>
        }
      },
      {
        alias: "$group_4"
        expr: {
          chain: [
            "$group_4"
          ]
          type: {
            name: "$group_4"
            table_type: <recursion ...>
          }
        }
        hidden: True
        type: {
          alias: "$group_4"
          type: <recursion ...>
        }
      }
    ]
    select_from: {
      table: {
        select: [
          {
            alias: "uuid"
            expr: {
              chain: [
                "uuid"
              ]
              type: <recursion ...>
            }
            hidden: True
            type: <recursion ...>
          },
          {
            alias: "event"
            expr: {
              chain: [
                "event"
              ]
              type: <recursion ...>
            }
            hidden: True
            type: <recursion ...>
          },
          {
            alias: "properties"
            expr: {
              chain: [
                "properties"
              ]
              type: <recursion ...>
            }
            hidden: True
            type: <recursion ...>
          },
          {
            alias: "timestamp"
            expr: {
              chain: [
                "timestamp"
              ]
              type: <recursion ...>
            }
            hidden: True
            type: <recursion ...>
          },
          {
            alias: "distinct_id"
            expr: {
              chain: [
                "distinct_id"
              ]
              type: <recursion ...>
            }
            hidden: True
            type: <recursion ...>
          },
          {
            alias: "elements_chain"
            expr: {
              chain: [
                "elements_chain"
              ]
              type: <recursion ...>
            }
            hidden: True
            type: <recursion ...>
          },
          {
            alias: "created_at"
            expr: {
              chain: [
                "created_at"
              ]
              type: <recursion ...>
            }
            hidden: True
            type: <recursion ...>
          },
          {
            alias: "$session_id"
            expr: {
              chain: [
                "$session_id"
              ]
              type: <recursion ...>
            }
            hidden: True
            type: <recursion ...>
          },
          {
            alias: "$group_0"
            expr: {
              chain: [
                "$group_0"
              ]
              type: <recursion ...>
            }
            hidden: True
            type: <recursion ...>
          },
          {
            alias: "$group_1"
            expr: {
              chain: [
                "$group_1"
              ]
              type: <recursion ...>
            }
            hidden: True
            type: <recursion ...>
          },
          {
            alias: "$group_2"
            expr: {
              chain: [
                "$group_2"
              ]
              type: <recursion ...>
            }
            hidden: True
            type: <recursion ...>
          },
          {
            alias: "$group_3"
            expr: {
              chain: [
                "$group_3"
              ]
              type: <recursion ...>
            }
            hidden: True
            type: <recursion ...>
          },
          {
            alias: "$group_4"
            expr: {
              chain: [
                "$group_4"
              ]
              type: <recursion ...>
            }
            hidden: True
            type: <recursion ...>
          }
        ]
        select_from: {
          table: {
            chain: [
              "events"
            ]
            type: <recursion ...>
          }
          type: <recursion ...>
        }
        type: <recursion ...>
      }
      type: <recursion ...>
    }
    type: {
      aliases: {}
      anonymous_tables: [
        <recursion ...>
      ]
      columns: {
        $group_0: <recursion ...>,
        $group_1: <recursion ...>,
        $group_2: <recursion ...>,
        $group_3: <recursion ...>,
        $group_4: <recursion ...>,
        $session_id: <recursion ...>,
        created_at: <recursion ...>,
        distinct_id: <recursion ...>,
        elements_chain: <recursion ...>,
        event: <recursion ...>,
        properties: <recursion ...>,
        timestamp: <recursion ...>,
        uuid: <recursion ...>
      }
      ctes: {}
      tables: {}
    }
  }
  '''
# ---
